Update CNG algorithm pseudo-handles article

denkisikinoneko opened the pull request:
Cleaned up redundant information and made the contents more consistent, and attached handle locations for ease of use in non-C++/C applications, from bcrypt.h (version 10.0.26100.0). Added links to IETF / NIST for each specification referenced (authoritative links for ANSI / PKCS could not be found). Some specifications linked may not match up with exactly what is implemented with Windows versions they were shipped with, and the most recent specified was used in place.

denkisikinoneko committed:
Update CNG algorithm pseudo-handles article

Cleaned up redundant information and made the contents more consistent, and attached handle locations for ease of use in non-C++/C applications, from bcrypt.h (version 10.0.26100.0). Added links to IETF / NIST for each specification referenced (authoritative links for ANSI / PKCS could not be found). Some specifications linked may not match up with exactly what is implemented with Windows versions they were shipped with, and the most recent specified was used in place.

denkisikinoneko commented:
@microsoft-github-policy-service agree

alvinashcraft commented:
Thanks for the PR. Since this PR updates so much of the content, I'm going to loop in someone from the product team to make sure everything looks OK to merge.

denkisikinoneko committed:
Remove RC4 repetition

denkisikinoneko commented:
I'm unsure about what to do about the supposed handle "BCRYPT_RNG_FIPS186_DSA_ALG_HANDLE," as I can't find it the latest SDK. Do you know if it is a preview like the digital signature algorithms (in which, a preview notice would probably be beneficial), or does it not exist at all?

alvinashcraft commented:
> I'm unsure about what to do about the supposed handle "BCRYPT_RNG_FIPS186_DSA_ALG_HANDLE," as I can't find it the latest SDK. Do you know if it is a preview like the digital signature algorithms (in which, a preview notice would probably be beneficial), or does it not exist at all?

I'm checking with the team about this handle. Thanks!

samuel-lee-msft commented on desktop-src/SecCNG/cng-algorithm-pseudo-handles.md:
```diff
@@ -12,119 +12,119 @@ Beginning in Windows 10, CNG (Cryptography API: Next Generation) provides pre-de
+| Constant | Description | Standards |
+|----------|-------------|-----------|
+| **BCRYPT_RNG_ALG_HANDLE**<br/>0x00000081 | A handle to the random-number generator algorithm. | **Windows 8:** The RNG algorithm supports [FIPS 186-3](https://csrc.nist.gov/files/pubs/fips/186-3/final/docs/fips_186-3.pdf). Keys less than or equal to 1024 bits adhere to [FIPS 186-2](https://csrc.nist.gov/files/pubs/fips/186-2/upd1/final/docs/fips186-2-change1.pdf) and keys greater than 1024 to [ ...
+| **BCRYPT_RNG_FIPS186_DSA_ALG_HANDLE** | A handle to the random-number generator algorithm suitable for DSA (Digital Signature Algorithm). | [FIPS 186-5](https://nvlpubs.nist.gov/nistpubs/FIPS/NIST.FIPS.186-5.pdf) |
```
Yup, pseudohandle should be removed as it has never existed.

Per [https://www.microsoft.com/en-us/security/blog/2019/11/25/going-in-depth-on-the-windows-10-random-number-generation-infrastructure/](https://www.microsoft.com/en-us/security/blog/2019/11/25/going-in-depth-on-the-windows-10-random-number-generation-infrastructure/) -> [https://aka.ms/win10rng](https://aka.ms/win10rng), the older `BCRYPT_RNG_FIPS186_DSA_ALGORITHM` and `BCRYPT_RNG_DUAL_EC_ALGORITHM` names are still supported but actually just route to AES-CTR-DRBGs, the same as `BCRYPT_RNG_ALGORITHM`.

We should update `desktop-src/SecCNG/cng-algorithm-identifiers.md` entry on `BCRYPT_RNG_FIPS186_DSA_ALGORITHM` to have the same text as `BCRYPT_RNG_DUAL_EC_ALGORITHM` in this respect:
"Windows 10: Beginning with Windows 10, the FIPS186 DSA random number generator algorithm has been removed. Existing uses of this algorithm will continue to work; however, the random number generator is based on the AES counter mode specified in the NIST SP 800-90 standard. New code should use BCRYPT_RNG_ALGORITHM, and it is recommended that existing code be changed to use BCRYPT_RNG_ALGORITHM."

denkisikinoneko replied:
Removed in commit 6b9547fa9be590b066473b6e5faf1219acfc2c02

denkisikinoneko replied:
Removed in commit 6b9547fa9be590b066473b6e5faf1219acfc2c02

samuel-lee-msft commented on desktop-src/SecCNG/cng-algorithm-pseudo-handles.md:
```diff
@@ -12,119 +12,119 @@ Beginning in Windows 10, CNG (Cryptography API: Next Generation) provides pre-de
+| **BCRYPT_HMAC_MD2_ALG_HANDLE**<br/>0x00000121 | A handle for the hash-based message authentication code using the MD2 hash algorithm. | [RFC 1319](https://www.ietf.org/rfc/rfc1319.txt), [RFC 2104](https://www.ietf.org/rfc/rfc2104.txt) |
+| **BCRYPT_HMAC_MD4_ALG_HANDLE**<br/>0x00000131 | A handle for the hash-based message authentication code using the MD4 hash algorithm. | [RFC 1320](https://www.ietf.org/rfc/rfc1320.txt), [RFC 2104](https://www.ietf.org/rfc/rfc2104.txt) |
+| **BCRYPT_HMAC_MD5_ALG_HANDLE**<br/>0x00000091 | A handle for the hash-based message authentication code using the MD4 hash algorithm. | [RFC 1321](https://www.ietf.org/rfc/rfc1321.txt), [RFC 2104](https://www.ietf.org/rfc/rfc2104.txt) |
+| Available in **Windows 11, version 23H2** |
```
>3 [](http://example.com/codeflow?start=38&length=1)

Is this true?
I think it might just be from Win11 24H2

denkisikinoneko replied:
The handles below are put under a guard in the header for NTDDI_VERSION >= NTDDI_WIN11_ZN, which, from what I could find on the Internet, is "Windows 10.0.25246-25398 / 23H2 / Zinc"
The same sources of information state 24H2 as "NTDDI_WIN11_GE" (with the latest handles in the header file being "NTDDI_WIN11_GA", 24H1)

samuel-lee-msft commented on desktop-src/SecCNG/cng-algorithm-pseudo-handles.md:
```diff
@@ -12,119 +12,119 @@ Beginning in Windows 10, CNG (Cryptography API: Next Generation) provides pre-de
-| **BCRYPT_SP80056A_CONCAT_ALG_HANDLE** | A handle to the SP800-56A key derivation function algorithm. Used by the [BCryptKeyDerivation](/windows/win32/api/Bcrypt/nf-bcrypt-bcryptkeyderivation) and [NCryptKeyDerivation](/windows/win32/api/Ncrypt/nf-ncrypt-ncryptkeyderivation) functions. |
+|----------|-------------|
+| **BCRYPT_CAPI_KDF_ALG_HANDLE**<br/>0x00000321 | A handle to the Crypto API (CAPI) key derivation function algorithm. Used by the [BCryptKeyDerivation](/windows/win32/api/Bcrypt/nf-bcrypt-bcryptkeyderivation) and [NCryptKeyDerivation](/windows/win32/api/Ncrypt/nf-ncrypt-ncryptkeyderivation) functions. |
+| **BCRYPT_PBKDF2_ALG_HANDLE**<br/>0x00000331 | A handle to the Password-based key derivation function 2 (PBKDF2) algorithm. Used by the [**BCryptKeyDerivation**](/windows/win32/api/Bcrypt/nf-bcrypt-bcryptkeyderivation) and [NCryptKeyDerivation](/windows/win32/api/Ncrypt/nf-ncrypt-ncryptkeyderivation) functions. |
```
>**BCryptKeyDerivation** [](http://example.com/codeflow?start=138&length=23)

Rogue bold

denkisikinoneko replied:
Removed in commit 6b9547fa9be590b066473b6e5faf1219acfc2c02

samuel-lee-msft commented on desktop-src/SecCNG/cng-algorithm-pseudo-handles.md:
```diff
@@ -12,119 +12,119 @@ Beginning in Windows 10, CNG (Cryptography API: Next Generation) provides pre-de
-| **BCRYPT_RNG_FIPS186_DSA_ALG_HANDLE** | A handle to the random-number generator algorithm suitable for DSA (Digital Signature Algorithm). Standard: FIPS 186-2. |
+| Constant | Description | Standards |
+|----------|-------------|-----------|
+| **BCRYPT_RNG_ALG_HANDLE**<br/>0x00000081 | A handle to the random-number generator algorithm. | **Windows 8:** The RNG algorithm supports [FIPS 186-3](https://csrc.nist.gov/files/pubs/fips/186-3/final/docs/fips_186-3.pdf). Keys less than or equal to 1024 bits adhere to [FIPS 186-2](https://csrc.nist.gov/files/pubs/fips/186-2/upd1/final/docs/fips186-2-change1.pdf) and keys greater than 1024 to [ ...
```
>**Windows 8:** The RNG algorithm supports [FIPS 186-3](https://csrc.nist.gov/files/pubs/fips/186-3/final/docs/fips_186-3.pdf). Keys less than or equal to 1024 bits adhere to [FIPS 186-2](https://csrc.nist.gov/files/pubs/fips/186-2/upd1/final/docs/fips186-2-change1.pdf) and keys greater than 1024 to [FIPS 186-3](https://csrc.nist.gov/files/pubs/fips/186-3/final/docs/fips_186-3.pdf).<br/> **Windows Vista with SP1 and Windows Server 2008:** The random number generator is based on the AES counter mode specified in the [NIST SP 800-90](https://nvlpubs.nist.gov/nistpubs/Legacy/SP/nistspecialpublication800-90r.pdf) standard.<br/> **Windows Vista:** The random number generator is based on the hash-based random number generator specified in the [FIPS 186-2](https://csrc.nist.gov/files/pubs/fips/186-2/upd1/final/docs/fips186-2-change1.pdf) standard [](http://example.com/codeflow?start=98&length=850)

The text is confused because Windows 10+ certainly are only AES-CTR-DRBG; would need to do some archaeology to understand when in the past other DRBGs were used but I don't think it's super relevant to this documentation.

Maybe just simplify to speak to Windows 10+, and the same choices should be made for cng-algorithm-identifiers.md

@alvinashcraft what do you think?

alvinashcraft replied:
I agree. Unless we have good reason, we try not to have notes about very old, unsupported versions. We can use language like "Starting in Windows 10,".

denkisikinoneko replied:
Updated in commit 594655752b4e6ca2e90386eff5495861527ebf49

samuel-lee-msft commented:
Left a few comments. Happy with the changes at a high level.

General comment that links are great, but also can become dead over time. Up to doc folks on best approach.
I don't think it's a huge burden on folks who are interested to find the underlying standards docs without the explicit links.

denkisikinoneko committed:
Remove BCRYPT_RNG_FIPS186_DSA_ALG_HANDLE, rouge bold

denkisikinoneko committed:
Use less likely to rot URIs

denkisikinoneko committed:
Specify Windows 10+ RNG algorithm only

Revise RNG description based off https://github.com/MicrosoftDocs/win32/pull/2124#discussion_r2411973465

alvinashcraft reviewed (commented):
This all looks much better to me. Let me know if you have any remaining concerns, @samuel-lee-msft. Otherwise, we can move ahead with the PR. Thanks folks!

alvinashcraft commented:
I'm merging this today, but I'm happy to circle back and make additional updates if anything needs to be tweaked later. Thanks for your contribution @denkisikinoneko

jess-krynitsky commented:
[like]  Jessica Krynitsky reacted to your message:
________________________________
From: Alvin Ashcraft ***@***.***>
Sent: Monday, October 20, 2025 8:22:59 PM
To: MicrosoftDocs/win32 ***@***.***>
Cc: Jessica Krynitsky ***@***.***>; Review requested ***@***.***>
Subject: Re: [MicrosoftDocs/win32] Update CNG algorithm pseudo-handles article (PR #2124)


Merged #2124<https://github.com/MicrosoftDocs/win32/pull/2124> into docs.

—
Reply to this email directly, view it on GitHub<https://github.com/MicrosoftDocs/win32/pull/2124#event-20388342499>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ASCZ5QUWKXBSJIA2MHOMKK33YVACHAVCNFSM6AAAAACIKIASFSVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMRQGM4DQMZUGI2DSOI>.
You are receiving this because your review was requested.Message ID: ***@***.***>
